lib/acpi/acpi.c

```diff
@@ -0,0 +1,623 @@
+		LOG_ERR("unable to retrieve IRQ Routing Table: %s", bus_name);
+		return -EIO;
+	}
+
+	for (int i = 0; i < CONFIG_ACPI_MAX_PRT_ENTRY; i++) {
+		if (!bus->pci_prt_table[i].SourceIndex) {
+			break;
+		}
+		/* mark the PRT irq numbers as reserved. */
+		arch_irq_set_used(bus->pci_prt_table[i].SourceIndex);
+	}
+
+	return 0;
+}
+
+static int acpi_retrieve_legacy_irq(struct acpi *bus)
+{
+	/* TODO: assume platform have only one PCH with single PCI bus (bus 0). */
+	return acpi_get_irq_table(bus, "_SB.PC00", bus->pci_prt_table, sizeof(bus->pci_prt_table));
+}
+
+static ACPI_STATUS dev_resource_enum_callback(ACPI_HANDLE obj_handle, UINT32 level, void *ctx,
+					      void **ret_value)
+{
+	ACPI_NAMESPACE_NODE *node;
+	ACPI_BUFFER rt_buffer;
+	struct acpi *bus = (struct acpi *)ctx;
+	struct acpi_dev *child_dev;
+
+	node = ACPI_CAST_PTR(ACPI_NAMESPACE_NODE, obj_handle);
+	char *path_name;
+	int status;
+	ACPI_DEVICE_INFO *dev_info;
+
+	LOG_DBG("%s %p\n", __func__, node);
+
+	/* get device info such as HID, Class ID etc. */
+	status = AcpiGetObjectInfo(obj_handle, &dev_info);
+	if (ACPI_FAILURE(status)) {
+		LOG_ERR("AcpiGetObjectInfo failed: %s", AcpiFormatException(status));
+		goto exit;
+	}
+
+	if (bus->num_dev >= CONFIG_ACPI_DEV_MAX) {
+		return AE_NO_MEMORY;
+	}
+
+	child_dev = (struct acpi_dev *)&bus->child_dev[bus->num_dev++];
+	child_dev->handle = obj_handle;
+	child_dev->dev_info = dev_info;
+
+	path_name = AcpiNsGetNormalizedPathname(node, TRUE);
+	if (!path_name) {
+		LOG_ERR("No memory for path_name\n");
+		goto exit;
+	} else {
+		LOG_DBG("Device path: %s\n", path_name);
+		child_dev->path = path_name;
+	}
+
+	rt_buffer.Pointer = NULL;
+	rt_buffer.Length = ACPI_ALLOCATE_LOCAL_BUFFER;
+
+	status = AcpiGetCurrentResources(node, &rt_buffer);
+	if (ACPI_FAILURE(status)) {
+		LOG_DBG("AcpiGetCurrentResources failed: %s", AcpiFormatException(status));
+	} else {
+		child_dev->res_lst = rt_buffer.Pointer;
+	}
+
+exit:
+
+	return status;
+}
+
+static int acpi_enum_devices(struct acpi *bus)
+{
+	LOG_DBG("");
+
+	AcpiWalkNamespace(ACPI_TYPE_DEVICE, ACPI_ROOT_OBJECT, ACPI_UINT32_MAX,
+			  dev_resource_enum_callback, NULL, bus, NULL);
+
+	return 0;
+}
+
+int acpi_init(void)
+{
+	int status;
+
+	LOG_DBG("");
+
+	if (bus_ctx.status != AE_NOT_CONFIGURED) {
+		LOG_DBG("acpi init already done");
+		return bus_ctx.status;
+	}
+
+	/* For debug version only */
+	ACPI_DEBUG_INITIALIZE();
+
+	status = initialize_acpica();
+	if (ACPI_FAILURE(status)) {
+		LOG_ERR("Error in ACPI init:%d", status);
+		goto exit;
+	}
+
+	/* Enable IO APIC mode */
+	status = acpi_enable_pic_mode();
+	if (ACPI_FAILURE(status)) {
+		LOG_WRN("Error in enable pic mode acpi method:%d", status);
+	}
+
+	status = acpi_retrieve_legacy_irq(&bus_ctx);
+	if (status) {
+		LOG_ERR("Error in retrieve legacy interrupt info:%d", status);
+		goto exit;
+	}
+
+	acpi_enum_devices(&bus_ctx);
+
+exit:
+	bus_ctx.status = status;
+
+	return status;
+}
+
+int acpi_early_init(void)
```

Comment:
Where is this function called from? I don't see any use of it in this PR.

Reply:
This API is useful for early platform init such as for CPU and currently it is given as placeholder

Reply:
Understood, however I think it would be better to add it only once you have a place that will call it, i.e. in some later PR. This will help explain its purpose as well as ensure proper code/testing coverage.

Reply:
Add shell command for retrieve acpi table and this will invoke acpi_early_init.

Reply:
The shell runs very late in the boot, and since you're entering commands it's pretty much guaranteed that everything happens *after* the boot. So I'm a bit confused by something called "early init" would be appropriate to be called from there.

Reply:
acpi_early_init  verified locally in pre-kernel stage locally.

Reply:
> acpi_early_init verified locally in pre-kernel stage locally.

Cannot understand this sentence.